### PR TITLE
Added support for RDS

### PR DESCRIPTION
**Pull request by fforloff**

### CODE REVIEW

**Comment by krutisfood** [comment]
Great addition & thanks for the PR @fforloff !

I put it in for testing, currently failing with ```lambdaRole/hammertime-rds-stop-hammertime is not authorized to perform: rds:DescribeDBInstances```, will have to update the IAM policy in serverless.yml, at a quick search rds:ListTagsForResource, rds:StartDBInstance, rds:StopDBInstance, rds:AddTagsToResource, rds:RemoveTagsFromResource should get it most of the way.

**Comment by fforloff** [comment]
Hi Kurt,

Looks like my update to serverless.yml gone missing when copying from the private repo. Yes, you are right, we would need those 6 rds actions you have mentioned.

Regards,

Victor

**Comment by fforloff** [comment]
Hi Kurt,
I've updated serverless.ymlk as discussed. Please review :-)

File: src/rds/filterStartableDBInstances.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+module.exports = function filterStartableDBInstances() {
+  return new Promise((resolve, reject) => {
+    listAllDBInstances()
+      .then(data => {
+        resolve(data
+          .DBInstances
+          .filter(allStartableDBInstances)
+          .map(instance => instance.DBInstanceArn));
+      })
+      .catch(err => reject(err));
```
Comment (ramesius) [inline src/rds/filterStartableDBInstances.js]: No need to wrap in a promise. Can also then remove the catch that seemingly just throw the exception again.

File: src/rds/filterStoppableDBInstances.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+module.exports = function filterStoppableDBInstances() {
+  return new Promise((resolve, reject) => {
+    listAllDBInstances()
+      .then(data => {
+        resolve(data
+          .DBInstances
+          .filter(allStoppableDBInstances)
+          .map(instance => instance.DBInstanceArn));
+      })
+      .catch(err => reject(err));
```
Comment (ramesius) [inline src/rds/filterStoppableDBInstances.js]: No need to wrap in a promise. Can also then remove the catch that seemingly just throw the exception again.

File: src/rds/listAllDBInstances.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const AWS = require('aws-sdk');
+
+module.exports = function listAllDBInstances() {
+  return new Promise((resolve, reject) => {
+    const rds = new AWS.RDS();
+    rds.describeDBInstances()
+      .promise()
+      .then(data => resolve(data))
+      .catch(reject);
```
Comment (ramesius) [inline src/rds/listAllDBInstances.js]: Unnecessary wrapped promise and catch.

File: src/rds/listDBInstancesToStart.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const filterStartableDBInstances = require('./filterStartableDBInstances');
+const filterDBInstancesTaggedToStart = require('./filterDBInstancesTaggedToStart');
+
+module.exports = function listDBInstancesToStart() {
+  return new Promise((resolve, reject) => {
+    filterStartableDBInstances()
+      .then(arns => filterDBInstancesTaggedToStart(arns))
+      .then(arns => resolve(arns))
+      .catch(reject);
```
Comment (ramesius) [inline src/rds/listDBInstancesToStart.js]: Unnecessary wrapped promise and catch.

File: src/rds/listDBInstancesToStop.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const filterStoppableDBInstances = require('./filterStoppableDBInstances');
+const filterDBInstancesTaggedUntouchable = require('./filterDBInstancesTaggedUntouchable');
+
+module.exports = function listDBInstancesToStop() {
+  return new Promise((resolve, reject) => {
+    filterStoppableDBInstances()
+      .then(arns => filterDBInstancesTaggedUntouchable(arns))
+      .then(arns => resolve(arns))
+      .catch(reject);
```
Comment (ramesius) [inline src/rds/listDBInstancesToStop.js]: Unnecessary wrapped promise and catch.

File: src/rds/notTaggedUntouchable.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+function notUntouchable(data) {
+  return !(data.TagList.some(tag => tag.Key === 'hammertime:canttouchthis'));
+};
+
+module.exports = function notTaggedUntouchable(arn) {
+  const params = {
+    ResourceName: arn
+  };
+  return new Promise((resolve, reject) => {
```
Comment (ramesius) [inline src/rds/notTaggedUntouchable.js]: Unnecessary wrapped promise and catch.

File: src/rds/startDBInstances.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+const startOneDBInstance = require('./startOneDBInstance');
+
+module.exports = function startDBInstances(arns) {
+  if (arns.length > 0) {
+    console.log("Starting RDS instances...");
+    const arnsMap = arns.map(arn => startOneDBInstance(arn));
+    return Promise.all(arnsMap);
+  } else {
+    console.log("No RDS instances to start... See you the next time...");
+    return [];
```
Comment (ramesius) [inline src/rds/startDBInstances.js]: Needs a `Promise.resolve([])' so return types are consistent.

File: src/rds/startOneDBInstance.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+const AWS = require('aws-sdk');
+
+module.exports = function startOneDBInstance(arn) {
+  const rds = new AWS.RDS();
+
+  var instanceId = arn.split(':').pop();
+  console.log("Starting "+instanceId+" ...");
+  return new Promise((resolve, reject) => {
```
Comment (ramesius) [inline src/rds/startOneDBInstance.js]: Unnecessary wrapped promise and catch.

File: src/rds/stopOneDBInstance.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+const AWS = require('aws-sdk');
+
+module.exports =function stopOneDBInstance(arn) {
+  const rds = new AWS.RDS();
+
+  var instanceId = arn.split(':').pop();
+  console.log("Stopping "+instanceId+"...");
+  return new Promise((resolve, reject) => {
```
Comment (ramesius) [inline src/rds/stopOneDBInstance.js]: Unnecessary wrapped promise and catch.

File: src/rds/tagOneDBInstance.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+module.exports = function tagOneDBInstance(arn) {
+  const params = {
+    ResourceName: arn,
+    Tags: [{
+      Key: 'hammertime:stop',
+      Value: new Date().toISOString()
+    }]
+  };
+  return new Promise((resolve, reject) => {
```
Comment (ramesius) [inline src/rds/tagOneDBInstance.js]: Unnecessary wrapped promise and catch.

File: src/rds/taggedHammertimeStop.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+function hammertimeStop(data) {
+  return (data.TagList.some(tag => tag.Key === 'hammertime:stop'));
+}
+
+module.exports = function taggedHammertimeStop(arn) {
+  const params = {
+    ResourceName: arn
+  };
+  return new Promise((resolve, reject) => {
```
Comment (ramesius) [inline src/rds/taggedHammertimeStop.js]: Unnecessary wrapped promise and catch.

File: src/rds/untagOneDBInstance.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+const AWS = require('aws-sdk');
+
+module.exports = function untagOneDBInstance(arn) {
+  const params = {
+    ResourceName: arn,
+    TagKeys: ['hammertime:stop']
+  };
+  return new Promise((resolve, reject) => {
```
Comment (ramesius) [inline src/rds/untagOneDBInstance.js]: Unnecessary wrapped promise and catch.

File: test/rds/stopOneDBInstance.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
+const AWS = require('aws-sdk-mock');
```
Comment (ramesius) [inline test/rds/stopOneDBInstance.js]: This file needs the `.test.js` extension

File: src/rds/startOneDBInstance.js (hunk 8 above)
Comment (fforloff) [reply src/rds/startOneDBInstance.js]: Updated per most of your comments, except the *One* functions - these need a promise wraparound, as they need to return promise, not arn. I can probably restructure the code to extract arns in startDBInstances, but I like the idea of arn-in arn-out...

Comment (ramesius) [reply src/rds/startOneDBInstance.js]: Agree with your comment of arn-in arn-out, however you can achieve this as follows without the need to wrap in a promise
```
function func(arn) {
 return promiseChain().then(() => arn);
}
```
https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Basic_Syntax

File: src/rds/startDBInstances.js (hunk 7 above)
Comment (fforloff) [reply src/rds/startDBInstances.js]: WIth 
"return Promise.resolve([]).all(arnsMap);"
I am getting type error "Promise.resolve(...).all is not a function",

Comment (ramesius) [reply src/rds/startDBInstances.js]: Apologies, line 11 should read something along the lines of `return Promise.resolve([]);`. This makes sure that each branch of the if statement returns a promise.
See: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve

Comment (fforloff) [reply src/rds/startDBInstances.js]: Thanks, ramesius. Silly me. Updated. Sorry for a delay- spent last 2 days on the planes..

**Comment by fforloff** [comment]
@krutisfood Hi Kurt, can you please approve my changes? Thanks and regards, Victor.

**Comment by krutisfood** [comment]
Hey @fforloff, sorry have been off-grid for the last month, I've pulled the code in for testing tonight as a solid sanity check & will aim to get back to you in next couple of days!

**Comment by fforloff** [comment]
Excellent! Thanks Kurt.

**Pull request merged**